rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Our API allows us to display relations for some particular work using a number of fields: 
+
+* `parts` is an array containing the children of the work (i.e. `parts` for work `A` would contain `A/1` and `A/2`)
+
+* `partOf` is an array containing the ancestors of the work (i.e. `partOf` for work `A/1/X` would contain `A` and `A/1`)
+
+* `precededBy` is an array containing the siblings of the work earlier in sort order (i.e. `precededBy` for work `A/2` would contain `A/1`)
+
+* `succeededBy` is an array containing the siblings of the work later in sort order (i.e. `succeededBy` for work `A/1/X` would contain `A/1/Y`)
+
+Note that the works contained within any of these relation fields are partial (only a core number of fields).
+
+## Problem
+
+Currently we calculate work relations dynamically on the server during a request cycle in the API, by querying Elasticsearch given some work path: we first fetch the original work by ID, then sequentially send a group of follow up queries for the relations (using the Elasticsearch multi search functionality). This approach has a number of drawbacks:
+
+* The API performs unnecessary work which could be done up front to provide faster response times.
+
+* It is not simple to extend this functionality to display relation data on the multi-work API endpoint, with it currently limited to the single work endpoint (the naive method of one extra request per work would result in a great number of Elasticsearch queries within a single incoming HTTP request).
+
+* The fact that the relations are not indexed up front means we can not use related data within search queries (such as if we wanted to include the title of the root of the collection for any descendent works, for example).
+
+* The use of paths for encoding the structure of the tree couples relations tightly to Calm. There is upcoming use cases for relations beyond just archives, and on a conceptual level the goal of the pipeline should be to generate data which is abstracted away from any particular source system.
+
+We therefore need a method to generate relation data on works within the pipeline, in advance of indexing for the API.
+
+## Enricher
+
+In order to include relation data on the work model itself it is clear there needs to be an additional stage of the pipeline which is able to query relationships between works, and include data from related works onto the work itself. This proposed name for this stage is the enricher.
+
+It is important that this stage takes place after the merger: if denormalisation happens before this stage the related data may be inconsistent with what exists on the work post merging. For this reason the enricher should exist near the end of the pipeline immediately preceding the ID minter.
```

Comment:
I'm a bit confused in this paragraph: what is "this stage"? Is that the enricher or the merger?

An example or explanation of how we'd end up with inconsistent data would help: I think it's because `precededBy` is a sorted list, and you can't combine that if you're merging two works together, but maybe I'm missing something.

It has to come after the merger, but is there any requirement for where it goes before/after the ID minter? I think it's position relative to the ID minter is mostly irrelevant (and maybe mention of that should be omitted).

Reply:
Assuming the enricher has no knowledge of what happens inside the merger, and the merger can mangle multiple work data in any way it wants (say it takes works with titles `A` and `B` and produces a work with title `A + B`). Say the enricher receives work `C` where one of its children is `A`, then if it is before the merging the work would contain the following data:

```
{
  "parts": [
    {
      "title": "A",
      ...
    },
  ],
  ...
}
```

where what we really want is:

```
{
  "parts": [
    {
      "title": "A + B",
      ...
    },
  ],
  ...

Reply:
@jamieparkinson theoretically how do you think this applies to images?

Should they be being created at this stage, or after the merging has happened?